### ISSUE TITLE
Pulumi does not correctly capture order/dependency for provider

### ISSUE THREAD

**Issue by JonathanFraser** [pulumi/pulumi#9944]
### What happened?

Having a resource depend on a provider output doesn't create the necessary dependency. I have an eks cluster, and kubernetes resources on that cluster where created in the same stack. As the provider is created as an output of the cluster, the provider is unwrapped with apply at attached to CustomResourceArgs for the resources. However, the resources aren't correctly parented to the cluster. 

### Steps to reproduce

Within a stack:

I create an EKS cluster and pull out the kubernetes provider.
I then create a suite of kubernetes resources (ConfigFile) using the provider.
This deploys correctly. However pulumi doesn't correctly establish a dependency and when it's deleted the cluster can be deleted first preventing the resources from ever being deleted and the only recourse is to wipe the state. 

`
var cluster = new Cluster("foo", new ClusterArgs{...});

cluster.Provider.Apply(provider => new ConfigFile("bar", new ConfigFileArgs{File="bar.yaml"}, new ComponentResourceOptions{Provider=>provider});
`

### Expected Behavior

pulumi understands there is a dependency between foo and bar. "bar" becomes a child of "foo"

### Actual Behavior

pulumi doesn't consider the provider as a dependency. 
"bar" is not a child of "foo"

### Versions used

  Pulumi" "3.*" 
  Pulumi.Aws" "5.*"
  Pulumi.Eks" "0.40.0"
  Pulumi.Kubernetes" "3.19.3"

### Additional context

I would expect any resource created with an apply context to be parented to the object responsible for the output, regardless of how that parameter is used. 

`
var foo = new Resource1("foo"); 
foo.SomeOutput.Apply(t => new Resource2("bar"));
`

should still establish a dependency between foo->bar, trying to optimize or ignore that is only going to cause unpredictable behaviour. 

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by Frassle** [comment]
> pulumi understands there is a dependency between foo and bar. "bar" becomes a child of "foo"

I agree on the dependency point. I suspect we're maybe losing the dependency information around the provider object itself. 

I don't think it should be automatically parented to it. Parenting should be fully in control of the user, all its for is to set how resources are grouped in displays.

> I would expect any resource created with an apply context to be parented to the object responsible for the output, regardless of how that parameter is used.
> `var foo = new Resource1("foo"); foo.SomeOutput.Apply(t => new Resource2("bar"));`
> should still establish a dependency between foo->bar, trying to optimize or ignore that is only going to cause unpredictable behaviour.

This probably isn't doable, not every language has the capability to support the idea of implicit context like this safely and we want to make sure that outputs behave as similarly as possible between the different SDK.s

**Comment by lukehoban** [comment]
> Then I know why the dependency on provider is being dropped. It's because CustomResourceOptions don't take Input<>, they take the raw value.

I think this is ultimately https://github.com/pulumi/pulumi/issues/7012.

It is also the reason why https://github.com/pulumi/pulumi-eks/issues/602 was opened - as unfortunately the `.provider` property is largely misleading, and it is better to create a provider within your program using the `.KubeConfig` output - that workaround is shown in the linked issue.

**Comment by JonathanFraser** [comment]
#7012 would ultimately fix it I guess as it would lift everything into the Output world. However, it will also make things more of a pain because accessors and methods become more cumbersome. It also doesn't stop resources from being created in an apply context which is apparently just broken and expected to be so. In that sense I guess this is a "works as designed" .
